### ISSUE TITLE
Javascript empty output

### ISSUE THREAD

**Issue by amir-s** [zhblue/hustoj#234]
Hello,

I was wondering if it is just me, or a problem with the judge_client.

running javascript always produces "wrong answer" and the `user.out` file is always empty. 

I do have `nodejs` installed (using `apt-get install nodejs`). Running "Ubuntu 16.04.3 x64".

**Comment by zhblue** [comment]
try use debug mode , collect log message to analyse the course of the problem
https://github.com/zhblue/hustoj/blob/master/wiki/Debug.md

**Comment by amir-s** [comment]
Thanks for updating the client. I pulled those changes, and installed a fresh version. But it still doesn't work.

I tried to debug it, first it returned the runtime error for syscall `20`. Then I added that syscall to `okcalls64.h`, recompiled the client and ran it again. It finished running, but produced a wrong answer since the output is empty.
I posed some logs in the details section below:

<details>

```
ifconfig|grep 'inet'|awk -F: '{printf $2}'|awk  '{printf $1}'
/bin/umount -f /home/judge/run1//proc
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f /home/judge/run1//dev 
umount: /home/judge/run1//dev: mountpoint not found
/bin/umount -f /home/judge/run1//lib 
umount: /home/judge/run1//lib: mountpoint not found
/bin/umount -f /home/judge/run1//lib64 
umount: /home/judge/run1//lib64: mountpoint not found
/bin/umount -f /home/judge/run1//etc/alternatives 
umount: /home/judge/run1//etc/alternatives: mountpoint not found
/bin/umount -f /home/judge/run1//usr 
umount: /home/judge/run1//usr: mountpoint not found
/bin/umount -f /home/judge/run1//bin 
umount: /home/judge/run1//bin: mountpoint not found
/bin/umount -f /home/judge/run1//proc 
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f bin usr lib lib64 etc/alternatives proc dev 
umount: bin: mountpoint not found
umount: usr: mountpoint not found
umount: lib: mountpoint not found
umount: lib64: mountpoint not found
umount: etc/alternatives: mountpoint not found
umount: /proc: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
umount: dev: mountpoint not found
/bin/umount -f /home/judge/run1//* 
umount: /home/judge/run1//log: not mounted
/bin/umount -f /home/judge/run1//log/* 
umount: /home/judge/run1//log/Main.js: not mounted
umount: /home/judge/run1//log/bin: not mounted
umount: /home/judge/run1//log/ce.txt: not mounted
umount: /home/judge/run1//log/data.in: not mounted
umount: /home/judge/run1//log/diff.out: not mounted
umount: /home/judge/run1//log/error.out: not mounted
umount: /home/judge/run1//log/lib: not mounted
umount: /home/judge/run1//log/lib64: not mounted
umount: /home/judge/run1//log/nodejs: not mounted
umount: /home/judge/run1//log/user.out: not mounted
umount: /home/judge/run1//log/usr: not mounted
/bin/umount -f /home/judge/run1//log/etc/alternatives 
umount: /home/judge/run1//log/etc/alternatives: mountpoint not found
/bin/rm -rf /home/judge/run1//log/*
mkdir /home/judge/run1//log/
mkdir: cannot create directory ‘/home/judge/run1//log/’: File exists
/bin/mv /home/judge/run1//* /home/judge/run1//log/
/bin/mv: cannot move '/home/judge/run1//log' to a subdirectory of itself, '/home/judge/run1//log/log'
/bin/mkdir -p /dev/shm/hustoj//home/judge/run1/
/bin/ln -s /dev/shm/hustoj//home/judge/run1/ /home/judge/
/bin/ln: failed to create symbolic link '/home/judge/run1': File exists
/bin/chown judge /dev/shm/hustoj//home/judge/run1/ 
chmod 755 /dev/shm/hustoj//home/judge/run1/ 
/bin/ln -s /home/judge/data /dev/shm/hustoj//home/judge/
/bin/ln: failed to create symbolic link '/dev/shm/hustoj//home/judge/data': File exists
Main=Main.jschown judge /home/judge/run1//Main.js
time: 3 mem: 320
nothing to do!
compile end!
status=0
/bin/umount -f bin usr lib lib64 etc/alternatives proc dev 2>&1 >/dev/null
umount: bin: mountpoint not found
umount: usr: mountpoint not found
umount: lib: mountpoint not found
umount: lib64: mountpoint not found
umount: etc/alternatives: mountpoint not found
umount: /proc: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
umount: dev: mountpoint not found
/bin/umount -f /home/judge/run1//* 2>&1 >/dev/null
umount: /home/judge/run1//Main.js: not mounted
umount: /home/judge/run1//ce.txt: not mounted
umount: /home/judge/run1//log: not mounted
/bin/umount -f /home/judge/run1//proc
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f /home/judge/run1//dev 
umount: /home/judge/run1//dev: mountpoint not found
/bin/umount -f /home/judge/run1//lib 
umount: /home/judge/run1//lib: mountpoint not found
/bin/umount -f /home/judge/run1//lib64 
umount: /home/judge/run1//lib64: mountpoint not found
/bin/umount -f /home/judge/run1//etc/alternatives 
umount: /home/judge/run1//etc/alternatives: mountpoint not found
/bin/umount -f /home/judge/run1//usr 
umount: /home/judge/run1//usr: mountpoint not found
/bin/umount -f /home/judge/run1//bin 
umount: /home/judge/run1//bin: mountpoint not found
/bin/umount -f /home/judge/run1//proc 
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f bin usr lib lib64 etc/alternatives proc dev 
umount: bin: mountpoint not found
umount: usr: mountpoint not found
umount: lib: mountpoint not found
umount: lib64: mountpoint not found
umount: etc/alternatives: mountpoint not found
umount: /proc: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
umount: dev: mountpoint not found
/bin/umount -f /home/judge/run1//* 
umount: /home/judge/run1//Main.js: not mounted
umount: /home/judge/run1//ce.txt: not mounted
umount: /home/judge/run1//log: not mounted
/bin/umount -f /home/judge/run1//log/* 
umount: /home/judge/run1//log/*: mountpoint not found
/bin/umount -f /home/judge/run1//log/etc/alternatives 
umount: /home/judge/run1//log/etc/alternatives: mountpoint not found
/bin/umount -f /home/judge/run1//proc
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f /home/judge/run1//dev 
umount: /home/judge/run1//dev: mountpoint not found
/bin/umount -f /home/judge/run1//lib 
umount: /home/judge/run1//lib: mountpoint not found
/bin/umount -f /home/judge/run1//lib64 
umount: /home/judge/run1//lib64: mountpoint not found
/bin/umount -f /home/judge/run1//etc/alternatives 
umount: /home/judge/run1//etc/alternatives: mountpoint not found
/bin/umount -f /home/judge/run1//usr 
umount: /home/judge/run1//usr: mountpoint not found
/bin/umount -f /home/judge/run1//bin 
umount: /home/judge/run1//bin: mountpoint not found
/bin/umount -f /home/judge/run1//proc 
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f bin usr lib lib64 etc/alternatives proc dev 
umount: bin: mountpoint not found
umount: usr: mountpoint not found
umount: lib: mountpoint not found
umount: lib64: mountpoint not found
umount: etc/alternatives: mountpoint not found
umount: /proc: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
umount: dev: mountpoint not found
/bin/umount -f /home/judge/run1//* 
umount: /home/judge/run1//Main.js: not mounted
umount: /home/judge/run1//ce.txt: not mounted
umount: /home/judge/run1//log: not mounted
/bin/umount -f /home/judge/run1//log/* 
umount: /home/judge/run1//log/*: mountpoint not found
/bin/umount -f /home/judge/run1//log/etc/alternatives 
umount: /home/judge/run1//log/etc/alternatives: mountpoint not found
/bin/mkdir /home/judge/run1//lib
/bin/mkdir /home/judge/run1//lib64
/bin/mkdir /home/judge/run1//bin
/bin/cp -a /lib/x86_64-linux-gnu /home/judge/run1//lib/
/bin/cp /lib64/* /home/judge/run1//lib64/
/bin/cp /bin/busybox /home/judge/run1//bin/
/bin/ln -s /bin/busybox /home/judge/run1//bin/sh
/bin/cp /bin/bash /home/judge/run1//bin/bash
/bin/mkdir -p /home/judge/run1//usr/lib /home/judge/run1//lib/i386-linux-gnu/
/bin/cp /lib/i386-linux-gnu/libz.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libz.so.*': No such file or directory
/bin/cp /usr/lib/i386-linux-gnu/libuv.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/usr/lib/i386-linux-gnu/libuv.so.*': No such file or directory
/bin/cp /usr/lib/i386-linux-gnu/libicui18n.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/usr/lib/i386-linux-gnu/libicui18n.so.*': No such file or directory
/bin/cp /usr/lib/i386-linux-gnu/libicuuc.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/usr/lib/i386-linux-gnu/libicuuc.so.*': No such file or directory
/bin/cp /usr/lib/i386-linux-gnu/libicudata.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/usr/lib/i386-linux-gnu/libicudata.so.*': No such file or directory
/bin/cp /usr/lib/i386-linux-gnu/libcares.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/usr/lib/i386-linux-gnu/libcares.so.*': No such file or directory
/bin/cp /usr/lib/libv8.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/usr/lib/libv8.so.*': No such file or directory
/bin/cp /lib/i386-linux-gnu/libssl.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libssl.so.*': No such file or directory
/bin/cp /lib/i386-linux-gnu/libcrypto.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libcrypto.so.*': No such file or directory
/bin/cp /lib/i386-linux-gnu/libdl.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libdl.so.*': No such file or directory
/bin/cp /lib/i386-linux-gnu/librt.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/librt.so.*': No such file or directory
/bin/cp /usr/lib/i386-linux-gnu/libstdc++.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/usr/lib/i386-linux-gnu/libstdc++.so.*': No such file or directory
/bin/cp /lib/i386-linux-gnu/libpthread.so.*  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libpthread.so.*': No such file or directory
/bin/cp /lib/i386-linux-gnu/libc.so.6  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libc.so.6': No such file or directory
/bin/cp /lib/i386-linux-gnu/libm.so.6  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libm.so.6': No such file or directory
/bin/cp /lib/i386-linux-gnu/libgcc_s.so.1  /home/judge/run1//lib/i386-linux-gnu/
/bin/cp: cannot stat '/lib/i386-linux-gnu/libgcc_s.so.1': No such file or directory
/bin/cp /lib/ld-linux.so.*  /home/judge/run1//lib/
/bin/cp: cannot stat '/lib/ld-linux.so.*': No such file or directory
/bin/mkdir -p /home/judge/run1//usr/lib /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/libz.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /usr/lib/x86_64-linux-gnu/libcares.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp: cannot stat '/usr/lib/x86_64-linux-gnu/libcares.so.*': No such file or directory
/bin/cp /usr/lib/libv8.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp: cannot stat '/usr/lib/libv8.so.*': No such file or directory
/bin/cp /lib/x86_64-linux-gnu/libssl.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/libcrypto.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/libdl.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/librt.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /usr/lib/x86_64-linux-gnu/libstdc++.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/libpthread.so.*  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/libc.so.6  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/libm.so.6  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib/x86_64-linux-gnu/libgcc_s.so.1  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /lib64/ld-linux-x86-64.so.2  /home/judge/run1//lib/x86_64-linux-gnu/
/bin/cp /usr/lib/x86_64-linux-gnu/libcares* /home/judge/run1//usr/lib/
/bin/cp: cannot stat '/usr/lib/x86_64-linux-gnu/libcares*': No such file or directory
/bin/cp /usr/bin/nodejs /home/judge/run1//
sample
sample
/bin/cp '/home/judge/data/1000/sample.in' /home/judge/run1//data.in
/bin/cp /home/judge/data/1000/*.dic /home/judge/run1//
/bin/cp: cannot stat '/home/judge/data/1000/*.dic': No such file or directory
init_call_counter:16
pid=3190 judging /home/judge/data/1000/sample.in
/bin/grep 'python2' Main.py
/bin/grep: Main.py: No such file or directory
echo '========[sample.out]========='>>diff.out
echo 'Expected						      |	Yours'>>diff.out
diff '/home/judge/data/1000/sample.out' user.out -y|head -100>>diff.out
echo '
=============================='>>diff.out
fail test /home/judge/data/1000/sample.in
test0
test0
/bin/cp '/home/judge/data/1000/test0.in' /home/judge/run1//data.in
/bin/cp /home/judge/data/1000/*.dic /home/judge/run1//
/bin/cp: cannot stat '/home/judge/data/1000/*.dic': No such file or directory
init_call_counter:16
pid=3203 judging /home/judge/data/1000/test0.in
/bin/grep 'python2' Main.py
/bin/grep: Main.py: No such file or directory
echo '========[test0.out]========='>>diff.out
echo 'Expected						      |	Yours'>>diff.out
diff '/home/judge/data/1000/test0.out' user.out -y|head -100>>diff.out
echo '
=============================='>>diff.out
fail test /home/judge/data/1000/test0.in
test1
test1
/bin/cp '/home/judge/data/1000/test1.in' /home/judge/run1//data.in
/bin/cp /home/judge/data/1000/*.dic /home/judge/run1//
/bin/cp: cannot stat '/home/judge/data/1000/*.dic': No such file or directory
init_call_counter:16
pid=3216 judging /home/judge/data/1000/test1.in
/bin/grep 'python2' Main.py
/bin/grep: Main.py: No such file or directory
echo '========[test1.out]========='>>diff.out
echo 'Expected						      |	Yours'>>diff.out
diff '/home/judge/data/1000/test1.out' user.out -y|head -100>>diff.out
echo '
=============================='>>diff.out
fail test /home/judge/data/1000/test1.in
add diff info of 1010..... 
/bin/umount -f /home/judge/run1//proc
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f /home/judge/run1//dev 
umount: /home/judge/run1//dev: mountpoint not found
/bin/umount -f /home/judge/run1//lib 
umount: /home/judge/run1//lib: not mounted
/bin/umount -f /home/judge/run1//lib64 
umount: /home/judge/run1//lib64: not mounted
/bin/umount -f /home/judge/run1//etc/alternatives 
umount: /home/judge/run1//etc/alternatives: mountpoint not found
/bin/umount -f /home/judge/run1//usr 
umount: /home/judge/run1//usr: not mounted
/bin/umount -f /home/judge/run1//bin 
umount: /home/judge/run1//bin: not mounted
/bin/umount -f /home/judge/run1//proc 
umount: /home/judge/run1//proc: mountpoint not found
/bin/umount -f bin usr lib lib64 etc/alternatives proc dev 
umount: bin: not mounted
umount: usr: not mounted
umount: lib: not mounted
umount: lib64: not mounted
umount: etc/alternatives: mountpoint not found
umount: /proc: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
umount: dev: mountpoint not found
/bin/umount -f /home/judge/run1//* 
umount: /home/judge/run1//Main.js: not mounted
umount: /home/judge/run1//bin: not mounted
umount: /home/judge/run1//ce.txt: not mounted
umount: /home/judge/run1//data.in: not mounted
umount: /home/judge/run1//diff.out: not mounted
umount: /home/judge/run1//error.out: not mounted
umount: /home/judge/run1//lib: not mounted
umount: /home/judge/run1//lib64: not mounted
umount: /home/judge/run1//log: not mounted
umount: /home/judge/run1//nodejs: not mounted
umount: /home/judge/run1//user.out: not mounted
umount: /home/judge/run1//usr: not mounted
/bin/umount -f /home/judge/run1//log/* 
umount: /home/judge/run1//log/*: mountpoint not found
/bin/umount -f /home/judge/run1//log/etc/alternatives 
umount: /home/judge/run1//log/etc/alternatives: mountpoint not found
/bin/rm -rf /home/judge/run1//log/*
mkdir /home/judge/run1//log/
mkdir: cannot create directory ‘/home/judge/run1//log/’: File exists
/bin/mv /home/judge/run1//* /home/judge/run1//log/
/bin/mv: cannot move '/home/judge/run1//log' to a subdirectory of itself, '/home/judge/run1//log/log'
result=6
```
```
cat /home/judge/run1/log/user.out  # this is empty
```
```
nodejs /home/judge/run1/log/Main.js < /home/judge/run1/log/data.in 
5
```

**Comment by amir-s** [comment]
There are bunch of missing folders and stuff when we copy the JS runtime, could it be because of that?

How one could figure out which libraries and folders are needed for a language?

**Comment by zhblue** [comment]
more update has been commit , but i used a 32bits ubuntu 16 ,so you might try add syscalls for 64bits by yourself.
and i used this code for the "A+B problem", and it works in my test machine.

```
var readline = require('readline');
process.stdin.setEncoding('utf-8');

var rl = readline.createInterface({input: process.stdin, output: process.stdout, prompt:''});

var solveMeFirst = (a, b) => a+b;

rl.on('line', function (data) {
    arr = data.split(' ');
    if (arr && arr.length==2) {
        c = solveMeFirst(+arr[0], +arr[1]);
        process.stdout.write('' + c + '\n');
    }
});

```

**Comment by zhblue** [comment]
the main cause of this bug is that the latest nodejs try to use the /dev/random and /dev/urandom
it worked mainly after i insert these 2 lines in judge_client.cc
https://github.com/zhblue/hustoj/blob/master/trunk/core/judge_client/judge_client.cc#L1658

**Comment by amir-s** [comment]
let me boot up a 32bit ubuntu machine and try again.
but it's not an issue with the syscalls I think, since it doesn't report any violation there.
it could be the missing stuff when we call `copy_js_runtime`. there are bunch of folders that are not being copied since they are missing.

what version of nodejs did you try it with?
mine:
```
nodejs -v
v4.2.6
```
Is there a way of figuring out which runtime libraries `nodejs` is trying to use on a 64bit machine?

**Comment by zhblue** [comment]
mine too,
nodejs -v
v4.2.6
try this command
```
 ldd `which nodejs`
```
it output these lines in my machine
```
        linux-gate.so.1 =>  (0xb7760000)
        libz.so.1 => /lib/i386-linux-gnu/libz.so.1 (0xb7725000)
        libuv.so.1 => /usr/lib/i386-linux-gnu/libuv.so.1 (0xb76fe000)
        librt.so.1 => /lib/i386-linux-gnu/librt.so.1 (0xb76f4000)
        libpthread.so.0 => /lib/i386-linux-gnu/libpthread.so.0 (0xb76d7000)
        libdl.so.2 => /lib/i386-linux-gnu/libdl.so.2 (0xb76d2000)
        libssl.so.1.0.0 => /lib/i386-linux-gnu/libssl.so.1.0.0 (0xb7668000)
        libcrypto.so.1.0.0 => /lib/i386-linux-gnu/libcrypto.so.1.0.0 (0xb747c000)
        libicui18n.so.55 => /usr/lib/i386-linux-gnu/libicui18n.so.55 (0xb7208000)
        libicuuc.so.55 => /usr/lib/i386-linux-gnu/libicuuc.so.55 (0xb7072000)
        libstdc++.so.6 => /usr/lib/i386-linux-gnu/libstdc++.so.6 (0xb6efb000)
        libm.so.6 => /lib/i386-linux-gnu/libm.so.6 (0xb6ea6000)
        libgcc_s.so.1 => /lib/i386-linux-gnu/libgcc_s.so.1 (0xb6e89000)
        libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb6cd2000)
        /lib/ld-linux.so.2 (0x800e8000)
        libicudata.so.55 => /usr/lib/i386-linux-gnu/libicudata.so.55 (0xb541a000)

```
it shows what dynamic libs are needed by nodejs

**Comment by amir-s** [comment]
I just tried the latest `master` in a 32bit machine. it worked 👍 
I'm interested to make it work for 64bit too. 

that 
```
ldd `which nodejs`
```
is exactly what I needed to start digging into this! thanks man!

It's strange that on 64b it doesn't really error out. It just leaves the output file empty.

By the way, when I submit the code through the web UI, it just stays, waiting for the judge result. but running the judge client manually in the server updates the status of the run.

I'll let you know if I could fix the 64bit issue.
Thanks for your help!

**Comment by zhblue** [comment]
after a debug test run all files is move to the log dir under the runX dir
if you copy a few files into the right location, for example:
```
cp /bin/bash run0/log/bin
cp /usr/bin/ldd run0/log/bin
cp /usr/bin/strace run0/log/bin
 # you might need to find the 64bits version in /lib/x86_64-linux-gnu
cp /lib/i386-linux-gnu/libtinfo.so.5 lib/i386-linux-gnu/   
```
you cant chroot into the log dir
```
chroot run0/log
```
and you can use strace to check out what nodejs is accually doing
```
 strace ./nodejs Main.js <data.in 2> log.txt
```
after that ,you can check log.txt line by line

**Comment by amir-s** [comment]
A lot of good information in your comment. Thanks! I really appreciate it!

I just used that `` ldd `which nodejs` `` and added the required DIRs to the `copy_js_runtime` function. Adding a few more syscalls to `okcalls64` and now it works on my 64bit server too!

I'll make a PR tomorrow after cleaning up my code!

**Comment by zhblue** [comment]
Congratulations!
Looking forward to your PR

**Comment by amir-s** [comment]
fixed via #236
